crates/core/src/error/collector.rs

```diff
@@ -0,0 +1,289 @@
+use derive_more::From;
+
+use crate::{error::RusticErrorKind, RusticError};
+use log::{error, info, warn};
+
+/// A rustic issue result
+///
+/// rustic issue results are used to return a result along with possible issues.
+#[derive(Debug)]
+pub enum RusticIssueResult<T> {
+    /// For when we return a result along with possible issues
+    Ok(T, Option<Vec<RusticIssue>>),
+
+    /// For when we abort and return only the errors that occurred
+    Err(Vec<RusticError>),
+}
+
+/// A rustic issue
+///
+/// An issue is a message that can be logged to the user.
+#[derive(Debug)]
+pub enum RusticIssue {
+    /// An error issue, indicating that something went wrong irrecoverably
+    Error(RusticError),
+
+    /// A warning issue, indicating that something might be wrong
+    Warning(RusticWarning),
+
+    /// An info issue, indicating additional information
+    Info(RusticInfo),
+}
+
+impl RusticIssue {
+    pub fn new_error(error: RusticErrorKind) -> Self {
+        Self::Error(error.into())
+    }
+
+    pub fn new_warning(message: &str) -> Self {
+        Self::Warning(message.into())
+    }
+
+    pub fn new_info(message: &str) -> Self {
+        Self::Info(message.into())
+    }
+
+    pub fn log(&self) {
+        match self {
+            Self::Error(error) => error!("{}", error),
+            Self::Warning(warning) => warn!("{}", warning.0),
+            Self::Info(info) => info!("{}", info.0),
+        }
+    }
+}
+
+/// A rustic warning message
+///
+/// Warning messages are used to indicate that something might be wrong.
+#[derive(Debug, Clone, From)]
+pub struct RusticWarning(String);
+
+impl RusticWarning {
+    pub fn new(message: &str) -> Self {
+        Self(message.to_owned())
+    }
+}
+
+impl From<&str> for RusticWarning {
+    fn from(message: &str) -> Self {
+        Self::new(message)
+    }
+}
+
+/// A rustic info message
+///
+/// Info messages are used to provide additional information to the user.
+#[derive(Debug, Clone, From)]
+pub struct RusticInfo(String);
+
+impl RusticInfo {
+    pub fn new(message: &str) -> Self {
+        Self(message.to_owned())
+    }
+}
+
+impl From<&str> for RusticInfo {
+    fn from(message: &str) -> Self {
+        Self::new(message)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct RusticIssueCollector {
+    /// The errors collected
+    errors: Option<Vec<RusticError>>,
+
+    /// The warnings collected
+    warnings: Option<Vec<RusticWarning>>,
+
+    /// The info collected
+    info: Option<Vec<RusticInfo>>,
+
+    /// Whether to log items directly during addition
+    log: bool,
+}
```

Comment:
We might want to adapt it, so we can easier use it from multiple threads, and other contexts, e.g. via Mutexes or Channels.